### PR TITLE
Feat/ddp

### PR DESCRIPTION
**Pull request by nom4dv3**
implement DataDestinationPlugin system.
Similar to DSP, it enables data destination supports in yaml in plugin style.
Difference: 
- DSP is defined on the combination of all the data source sections in the yaml file, so there's only 1 active combined DSP per yaml
- DDP is defined on the single section of the data destination section in the yaml file, so it would use N DDPs per yaml when it contains N data destination sections. 
- The execution order on multiple DDP case is corresponding to the order that the sections are defined in the yaml.

### CODE REVIEW

File: src/api/trigger.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { ddpHub } from '../ddp/hub'
+import type { CLEExecutable, ProofParams } from '../types'
+
+export async function trigger(
+  cleExecutable: Omit<CLEExecutable, 'wasmUint8Array'>,
+  cleId: string,
+  proofParams: ProofParams,
+  ddpParamsList: Record<string, any>[], // 1 ddpParams per ddp
+  _enableLog = false,
```
Comment (murongg) [inline src/api/trigger.ts:9]: can remove `_enableLog `

File: src/api/trigger.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+) {
+  const { cleYaml } = cleExecutable
+  const ddps = ddpHub.getDDPsByYaml(cleYaml)
+  for (const ddp of ddps)
+    await ddp.go(cleId, proofParams, ddpParamsList).catch((err) => { throw err })
```
Comment (murongg) [inline src/api/trigger.ts:14]: `catch` is unnecessary.

File: src/ddp/ethereum/index.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+export class EthereumDataDestinationPlugin extends DataDestinationPlugin<EthereumDDPGoParams> {
+  goParams: KeyofToArray<EthereumDDPGoParams> = ['signer']
+
+  async go(cleId: string, proofParams: ProofParams, goParams: EthereumDDPGoParams): Promise<void> {
+    const proof = ZkWasmUtil.bytesToBigIntArray(proofParams.aggregate_proof)
+    const instances = ZkWasmUtil.bytesToBigIntArray(proofParams.batch_instances)
+    const aux = ZkWasmUtil.bytesToBigIntArray(proofParams.aux)
+    const arg = ZkWasmUtil.bytesToBigIntArray(proofParams.instances)
+
+    try {
```
Comment (murongg) [inline src/ddp/ethereum/index.ts:22]: Do not need to throw an exception here?

Comment (nom4dv3) [reply src/ddp/ethereum/index.ts:22]: ah right, we do need, will update

**Review by lightmanM — APPROVED** [review]

File: src/ddp/ethereum/index.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      const graph = new ethers.Contract(cleId, graph_abi, goParams.signer)
+
+      const tx = await graph.trigger(proof, instances, aux, [arg], [0n], { gasLimit: goParams.gasLimit })
```
Comment (lightmanM) [inline src/ddp/ethereum/index.ts:25]: Should print tx hash and the receipt information with enablelog option

File: src/ddp/interface.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import type { KeyofToArray } from '@murongg/utils/index'
+import type { ProofParams } from '../types'
+import { dspParamsNormalize } from '../common/utils'
+
+export abstract class DataDestinationPlugin<GP extends object> {
+  abstract goParams: KeyofToArray<GP>
+  toGoParams(params: Record<string, any>) {
+    return dspParamsNormalize(this.goParams as string[], params) as GP
+  }
+  abstract go(cleId: string, proofParams: ProofParams, goParams: GP): Promise<void >
```
Comment (lightmanM) [inline src/ddp/interface.ts:10]: Is `go` really a good name for our case?

Comment (nom4dv3) [reply src/ddp/interface.ts:10]: it's like "go to destination". but do you have other suggestions?

File: src/ddp/ethereum/index.ts (hunk 4 above)
Comment (nom4dv3) [reply src/ddp/ethereum/index.ts:25]: eslint prevent from console statement, will do this in future commit

File: src/ddp/interface.ts (hunk 5 above)
Comment (nom4dv3) [reply src/ddp/interface.ts]: merging, we can still rename this later, let's keep this conv until decide

**Pull request merged**